packages/lisk-transaction-pool/src/queue.ts

```diff
@@ -22,18 +32,67 @@ export class Queue {
 		this._index = {};
 	}
 
+	public dequeueUntil(condition: (transaction: Transaction) => boolean): ReadonlyArray<Transaction> {
+		const { effected, uneffected } = this._transactions.reduceRight(({effected, uneffected, conditionFailedOnce}: dequeueUntilReduceObject, transaction: Transaction) => {
+			// Add transaction to the uneffected list if the condition failed for this transaction or any previous transaction
+			if (conditionFailedOnce || !condition(transaction)) {
+				return {
+					effected,
+					uneffected: [transaction, ...uneffected],
+					conditionFailedOnce: true,
+				};
+			}
+
+			// delete the index of the transaction which passed the condition
+			delete this._index[transaction.id];
+
+			return {
+				effected: [transaction, ...effected],
+				uneffected,
+				conditionFailedOnce: false,
+			};
+		}, {
+			effected: [],
+			uneffected: [],
+			conditionFailedOnce: false
+		});
+
+		this._transactions = uneffected;
+
+		return effected;
+	}
+
 	public enqueueMany(transactions: ReadonlyArray<Transaction>): void {
+		this._transactions = [...transactions, ...this._transactions];
+
+		transactions.forEach((transaction: Transaction) => {
+			this._index[transaction.id] = transaction;
+		});
 	}
 
 	public enqueueOne(transaction: Transaction): void {
+		this._transactions = [transaction, ...this._transactions];
+		this._index[transaction.id] = transaction;
 	}
 
 	public exists(transaction: Transaction): boolean {
+		return !!this._index[transaction.id];
 	}
 
 	public removeFor(condition: (transaction: Transaction) => boolean): ReadonlyArray<Transaction> {
```

Comment:
maybe it's good to define the `condition` function outside

Reply:
i am using condition when we are traversing through the transaction list. I think it's okay if we are just checking the hash object.